honeybadgermpc/protocols/AVID.py

```diff
@@ -0,0 +1,140 @@
+            m_list = input_list  # block until an input list is received
+            assert len(m_list) == self.n
+
+            # construct the stripe list which contains the stripes for party 1 to party n
+            stripes_list = [None] * self.n
+            for i in range(len(m_list)):
+                m = m_list[i]
+                stripes_list[i] = encode(k, self.n, m)
+
+            for i in range(self.n):
+                # send each person the column of stripes
+                stripes_for_each = [stripes[i] for stripes in stripes_list]
+                self.send(i, (sid, 'VAL', stripes_for_each))
+
+        result = [None] * self.n  # the result current party got from all parties
+        echo_set = set()
+        ready_set = set()
+        ready_sent = False
+        from_leader = None
+
+        def decode_output():
+            m = decode(k, self.n, result)
+            return m
+
+        while True:  # main receive loop
+            sender, msg = await self.receive()
+            if msg[1] == 'VAL' and from_leader is None:
+                # Validation
+                # TODO merkle tree validation
+                (_, _, stripes_for_each) = msg
+                if sender != self.leader:
+                    logger.info(f"[{pid}] VAL message from other than leader: {sender}")
+                    continue
+
+                # Update
+                from_leader = pid
+                # only send out the part each party need
+                # this is efficient by not sending the whole m
+                for i in range(self.n):
+                    self.send(i, (sid, 'ECHO', stripes_for_each[i]))
+
+            elif msg[1] == 'ECHO':
+                (_, _, stripe) = msg
+                # Validation
+                if result[sender] is not None or sender in echo_set:
+                    logger.info("[{pid}] Redundant ECHO")
+                    continue
+
+                # Update and set the result
+                result[sender] = stripe
+                echo_set.add(sender)
+
+                if len(echo_set) >= echo_threshold and not ready_sent:
+                    ready_sent = True
+                    broadcast((sid, 'READY'))
+
+            elif msg[1] == 'READY':
+                # Validation
+                if sender in ready_set:
+                    logger.info("[{pid}] Redundant READY")
+                    continue
+
+                # Update
+                ready_set.add(sender)
+
+                # Amplify ready messages
+                if len(ready_set) >= ready_threshold and not ready_sent:
+                    ready_sent = True
+                    broadcast((sid, 'READY'))
+
+                if len(ready_set) >= output_threshold and len(echo_set) >= k:
+                    # store the data inside the body
+                    self.__data[pid] = decode_output()
+                    broadcast((sid, 'STORE'))
+                    return
```

Comment:
Add a log in all places where you break out of the loop, this helps in debugging because you know that you are not stuck in the loop when you look at the log statement.

Reply:
resolved